test/test_torch.py

```diff
@@ -13395,6 +13395,35 @@ def test_cat_big(self, device):
         result = torch.cat(concat_list)
         self.assertEqual(result.size(0), SIZE1 + SIZE2)
 
+    @onlyCUDA
+    def test_atomic_add(self, device):
+        # todo: we should probably expose a dtype.is_signed()
+        def is_signed(dtype):
+            return torch.is_signed(torch.tensor(0, dtype=dtype))
+        # https://github.com/pytorch/pytorch/issues/29153
+        for dtype in torch.testing.get_all_math_dtypes(device):
+            size = [5, 5]
+            if dtype.is_floating_point:
+                tensor = torch.rand(size, dtype=dtype, device=device)
+            if is_signed(dtype):
+                tensor = torch.randint(-5, 15, size, dtype=dtype, device=device)
+            else:
+                tensor = torch.randint(0, 10, size, dtype=dtype, device=device)
```

Comment:
The atomicAdd bug looks like it is triggered in a very specific situation, when the sum being stored back into a memory address is signed. Would it be worth testing that case specifically?

Reply:
It needs to handle sign properly, but it wasn't just that: the original code just always returned zero. In intermediate versions I thought I fixed the issue, but I ran into issues that only occurred at specific alignments of values within the 32 bit words. In practice I don't feel like I got good test signal when I wrote specific narrowly defined tests for particular cases, and the randint for the signed case should always get some negatives.

Reply:
Could you give some sample inputs to index_add that trigger the bug? I want to know roughly the order of magnitude of the probability is that the randomly generated input does not catch the bug; we want to make our tests robust to changes in seed.

Reply:
In the example from the initial issue we actually got all zeros, so almost any random value would catch that. There are lots of possible ways this can be wrong though that were missed when I just tried to validate narrowly for particular values, and random tensors capture that more thoroughly.

Reply:
Could you walk through an example input and how it gets to all zeros? There's a lot of bit shifting going on and I'd like to make sure I understand how the implementation works

Reply:
I mean, nothing gets added at all. I'm not actually sure if there's a case that doesn't trigger this bug besides maybe adding a tensor of zeros. Here's an example for a single value with the expected result for int64 and the zero result for int16, it's basically a no-op:

```
# expected 0 + 1 == 1
>>> dtype=torch.int64
>>> torch.zeros([1], device='cuda', dtype=dtype).index_add(0, torch.tensor([0], device='cuda'), torch.ones([1], dtype=dtype, device='cuda'))
tensor([1], device='cuda:0')
>>> dtype=torch.int16
# unexpected 0 + 1 == 0
>>> torch.zeros([1], device='cuda', dtype=dtype).index_add(0, torch.tensor([0], device='cuda'), torch.ones([1], dtype=dtype, device='cuda'))
tensor([0], device='cuda:0', dtype=torch.int16)
```

It would be easy to add this example that motivated the change, but it's just so trivial that it doesn't add coverage. The test with a full multi-dimensional tensor and a variety of values can capture potential corner cases that this narrow test wouldn't.

Reply:
I'm not complaining about the test case coverage, I'm still trying to understand why the previous code is wrong. Because the fix was to change from doing unsigned math to signed math, there should be a non-trivial set of positive values for which the math is correct, but this doesn't seem to be the case from your explanation.

To help bridge my gap in understanding, it would be really helpful if you could run through an end-to-end example about how almost all inputs give zeros.

Reply:
I'm not sure it's worth trying to dig too deeply into all the ways the existing code was incorrect. It didn't work in any situation I know of and was never tested. There's more wrong with it than just the sign issue.

One thing that's wrong is that in `sum = val + (size_t)address & 2 ? T(old >> 16) : T(old & 0xffff);`, ... the `val + (size_t)address &2` as an expression has precedence. If val is non-zero, we'll always call `T(old>>16)` regardless of what the address was. Just fixing that, we still have the sign issue, but I'm not sure that completely covers the problems here.

There are so many ways untested code can be wrong, it's hard to be certain of all of them without a lot of investigation and additional testing.